File: DevEnvironment/Command/src/main/java/command/core/ProcessInformation.java
package command.core;

import base.core.Conditions;
import command.interfaces.IProcessInformation;
import java.time.Duration;
import java.time.Instant;
import java.util.Optional;

/**
 * The ProcessInformation class implements an information of a process.
 */
public final class ProcessInformation implements IProcessInformation {
    private final ProcessHandle processHandle;

    /**
     * Gets the current process information.
     */
    public static IProcessInformation current() {
        return new ProcessInformation(ProcessHandle.current());
    }

    /**
     * The ProcessInformation constructor.
     */
    public ProcessInformation(ProcessHandle processHandle) {
        Conditions.validateNotNull(
            processHandle,
            "The process handle.");

        this.processHandle = processHandle;
    }

    /**
     * Gets the processId.
     */
    @Override
    public long processId() {
        return this.processHandle.pid();
    }

    /**
     * Gets the user of the process.
     */
    @Override
    public Optional<String> user() {
        return this.processHandle.info().user();
    }

    /**
     * Gets the arguments of the process.
     */
    @Override
    public Optional<String[]> arguments() {
        return this.processHandle.info().arguments();
    }

    /**
     * Gets the executable pathname of the process.
     */
    @Override
    public Optional<String> command() {
        return this.processHandle.info().command();
    }

    /**
     * Gets the command line of the process.
     */
    @Override
    public Optional<String> commandLine() {
        return this.processHandle.info().commandLine();
    }

    /**
     * Gets the start time of the process.
     */
    @Override
    public Optional<Instant> startTime() {
        return this.processHandle.info().startInstant();
    }

    /**
     * Gets the total cpu time of the process.
     */
    @Override
    public Optional<Duration> totalCpuTime() {
        return this.processHandle.info().totalCpuDuration();
    }
}

File: SystemCore/Command/core/ProcessInformation.cpp
#include "ProcessInformation.h"

using namespace Command;

/**
 * The ProcessInformation constructor.
 */
ProcessInformation::ProcessInformation(
    unsigned int processId,
    const std::wstring& user,
    const std::vector<std::wstring>& arguments,
    const std::wstring& command,
    const std::wstring& commandLine,
    const base::DateTime& startTime,
    const base::Duration& totalCpuTime) :
    m_processId(processId),
    m_user(user),
    m_arguments(arguments),
    m_command(command),
    m_commandLine(commandLine),
    m_startTime(startTime),
    m_totalCpuTime(totalCpuTime)
{
}

/**
 * The ProcessInformation destructor.
 */
ProcessInformation::~ProcessInformation()
{
}

/**
 * Gets the ProcessId.
 */
unsigned int ProcessInformation::ProcessId() const
{
    return m_processId;
}

/**
 * Gets the User of the process.
 */
const std::wstring& ProcessInformation::User() const
{
    return m_user;
}

/**
 * Gets the Arguments of the process.
 */
void ProcessInformation::Arguments(std::vector<std::wstring>& arguments) const
{
    arguments.insert(arguments.end(), m_arguments.begin(), m_arguments.end());
}

/**
 * Gets the executable pathname of the process.
 */
const std::wstring& ProcessInformation::Command() const
{
    return m_command;
}

/**
 * Gets the Command line of the process.
 */
const std::wstring& ProcessInformation::CommandLine() const
{
    return m_commandLine;
}

/**
 * Gets the start time of the process.
 */
const base::DateTime& ProcessInformation::StartTime() const
{
    return m_startTime;
}

/**
 * Gets the total cpu time of the process.
 */
const base::Duration& ProcessInformation::TotalCpuTime() const
{
    return m_totalCpuTime;
}
